### ISSUE TITLE
Perform entire pageview update in one transaction

### ISSUE THREAD

**Issue by audiodude** [openzim/wp1#761]
We'd like to keep all of the cumulative pageviews in one column per article. However, this poses a problem if a update process gets interrupted, because we wouldn't know which rows have already been incremented with this month's data.

The solution is to do the entire update as one transaction, so that if the database updates, it means that all rows were successfully read and processed.

**Comment by ViSiOnOp19cr** [comment]
is this still open ??

**Comment by audiodude** [comment]
Looks like it. I think the current logic is still committing after every 50,000 rows.

https://github.com/openzim/wp1/blob/main/wp1/scores.py#L200

Note that the pageview data has not been calculated in production yet, so this code (and the update addressed by this issue) is currently unused.

**Comment by Aditi-Pande** [comment]
Hello everyone,

I would love to contribute to this issue! I have experience working with Python and backend development, and I’m excited to work on improving the Simple Selection Builder to extend URL stripping for non-English Wikipedia projects.

My approach:
1. I would first analyze the current URL stripping logic and identify where the logic is currently hardcoded for English wikipedia.
2. I would detect the project language dynamically by extracting the Wikipedia domain from the URL.
3. Generalize the stripping logic to handle all Wikipedia language versions
4. Ensure compatibility with both /wiki/ and /w/index.php?title= URL formats.

Would love any feedback or suggestions before I start! Let me know if there are any specific edge cases or considerations I should keep in mind. Looking forward to contributing! 

Regards,
Aditi Pande

**Comment by audiodude** [comment]
Hi @Aditi-Pande. I think you posted your message on the wrong issue.

**Comment by harsh-0409** [comment]
Hey @audiodude ,

I would like to contribute to this issue. I have come up with the following approach -:

Explicit Transaction Handling: Start a transaction explicitly using wp10db.begin() before making updates to ensure all modifications are treated as a single unit.
Error Handling & Rollback: If an error occurs at any point, roll back the transaction to maintain data integrity and prevent any inconsistent state.
Batch Commit Strategy: Instead of committing after every update, commit changes after processing a certain number of rows (commit_after) to optimize performance and reduce overhead.



Let me know if you have any suggestions.

**Comment by audiodude** [comment]
Yes that sounds right. We don't assign issues, but feel free to send a PR

**Comment by benoit74** [comment]
How many row updates are we speaking about? 

Beware that modifications done in a transaction and not yet committed to the DB are kept in a special area in the DB (called the undo log) which is limited in space (1G by default). If you make to much modifications in a single transaction, at some point this space will be exhausted and your transaction will rollback.

1G is still quite a lot, but this needs to be checked.

**Comment by harsh-0409** [comment]
Hey @audiodude ,

I have raise a PR for the issue. Kindly review.
